### PR TITLE
german parse overhaul

### PR DESCRIPTION
**Pull request by emphasize**
additions:
* german parse pretty much relied on a generic number parser which didn't account for spoken fractions, ordinals, decimals and so on, which made it a hit and miss expecially in date or duration parsing 
* rework of extract_number / extract_numbers and helpers
* german normalizer
* expansion of the test cases

refactor:
* num to string (and vice versa) dictionaries in common_data_de that also account for grammar

while the most work is in parse, format is also slightly affected

Both are positively tested.

------

This makes it possible to disentangle the draft `feat/timespan` from additional stuff that i brought in to solely focus on timespan.
I will update the branch accordingly  (when this PR is merged)

### CODE REVIEW

File: lingua_franca/lang/parse_de.py [hunk 1]
```diff
@@ -900,107 +1032,91 @@ def date_found():
...
-    if lowerstr[-2:] == "te":  # below 20 suffix is -te*
-        lowerstr = lowerstr[:-2]
-        if lowerstr in _DE_NUMBERS:
-            return _DE_NUMBERS[lowerstr]
 
-    if lowerstr[-3:] in ["ter", "tes", "ten", "tem"]:
-        lowerstr = lowerstr[:-3]
-        if lowerstr in _DE_NUMBERS:
-            return _DE_NUMBERS[lowerstr]
+def get_ordinal_index(input_str : str, type_: type = str):
```
Comment (JarbasAl) [inline lingua_franca/lang/parse_de.py:1098]: can these functions that are not shared across languages start with an underscore? `def _get_ordinal_index`

Comment (JarbasAl) [reply lingua_franca/lang/parse_de.py:1098]: or if you think some method is useful it can be added as a new function to be implemented per language

**Comment by JarbasAl** [comment]
I love this! if some native speaker can validate this and maybe suggest new cases for unittests it would be nice, if noone has any feedback I'm happy to merge this in a couple days

**Review by JarbasAl — APPROVED** [review]

**Comment by emphasize** [comment]
Still bringing in changes for the case `fractions=False` in `_convert_words_to_numbers_de()` . This will include tests (not much, but..)

**Comment by ChanceNCounter** [comment]
> Doing this i wonder why `extract_number`, `extract_numbers` don't have the option to disregard fractions?

It hadn't occurred to me. Can't speak for the others. My gut says "just do `int(extract_number(foo))` but I'm sure there are situations where that isn't the desired result. Anything come to mind?

**Comment by ChanceNCounter** [comment]
@thorstenMueller dreadfully sorry to bug your inbox, but I'd sure appreciate it if a native speaker could skim the resource files (at minimum) and you naturally came to mind

**Comment by thorstenMueller** [comment]
That's absolutely no problem, dear @ChanceNCounter . I'll give the changed files a quick check and report back to you.
Btw. i've used https://github.com/repodiac/german_transliterate previously.

**Comment by emphasize** [comment]
we should port some of `german_transliterate` logic (got my eye on that one too, backlogged), at least what fits in/on top of the normalizer.

**Comment by emphasize** [comment]
> Anything come to mind?

Both `extract_number` / `extract_numbers` call the `_extract_number_with_text` [#](https://github.com/emphasize/lingua-franca/blob/0f919343cd2f021c86b9bc5a1ecbbb30af5260b1/lingua_franca/lang/parse_en.py#L175) / `_extract_numbers_with_text` [#](https://github.com/emphasize/lingua-franca/blob/0f919343cd2f021c86b9bc5a1ecbbb30af5260b1/lingua_franca/lang/parse_en.py#L133) helper methods, which itself accept the `fractional_numbers` argument. Just need to add to their signature, if i'm not mistaken.

**Comment by emphasize** [comment]
Any suggestions before I merge?

File: lingua_franca/res/text/de-de/normalize.json [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      "fünfzehn": "15",
+      "sechzehn": "16",
+      "siebzehn": "17",
+      "achtzehn": "18",
+      "neunzehn": "19",
+      "zwanzig": "20",
+      "einundzwanzig": "21",
+      "zweiundzwanzig": "22",
+      "dreiundzwanzig": "23",
+      "vierundzwanzig": "24",
```
Comment (thorstenMueller) [inline lingua_franca/res/text/de-de/normalize.json:55]: What about 25-29?

File: test/unittests/test_parse_de.py [hunk 3]
```diff
@@ -33,55 +37,218 @@ def tearDownModule():
...
+            "es ist 7 8 9 test")
+        self.assertEqual(
+            normalize("es ist sieben acht neun test", lang="de-de"),
+            "es ist 7 8 9 test")
+        self.assertEqual(
+            normalize("dies ist zehn elf zwölf test", lang="de-de"),
+            "dies ist 10 11 12 test")
         self.assertEqual(
-            extract_number("dies ist der dritte Test", lang="de-de"), 3)
+            normalize("dies ist dreizehn vierzehn test", lang="de-de"),
```
Comment (thorstenMueller) [inline test/unittests/test_parse_de.py:75]: "dies ist dreizehn vierzehn test" isn't a really german sentence. Is this meant to be grammatically wrong? I'd say: "dies ist der dreizehnte vierzehnte test".

File: test/unittests/test_parse_de.py [hunk 4]
```diff
@@ -33,55 +37,218 @@ def tearDownModule():
...
+
+        self.assertEqual(extract_number("frisörtermin in 1 stunde"), 1)
+
+        self.assertEqual(extract_number("minus dreißig grad"), -30)
+
+        self.assertEqual(extract_number("das macht null sinn"), 0)
+
+        self.assertEqual(extract_number("eins punkt drei vier"), 1.34)
+
+        self.assertEqual(extract_number("eins komma drei vier"), 1.34)
```
Comment (thorstenMueller) [inline test/unittests/test_parse_de.py:120]: Is this converted to a dotted notation, not 1,34?

**Comment by thorstenMueller** [comment]
Sorry for not responding so long, i've been really busy lately. I've added some comments. Is this way of feedback meant by you?

File: test/unittests/test_parse_de.py (hunk 4 above)
Comment (emphasize) [reply test/unittests/test_parse_de.py:120]: this is parsing, not formatting. And "1,34" is not a number to work with.

Those two test should check if the point and comma utterance conversion works (and produces x.x)

File: test/unittests/test_parse_de.py (hunk 3 above)
Comment (emphasize) [reply test/unittests/test_parse_de.py:75]: i could have written "dreizehn vierzehn mal falsch geschrieben" :P. The normalizer doesn't deal with ordinals

File: lingua_franca/res/text/de-de/normalize.json (hunk 2 above)
Comment (emphasize) [reply lingua_franca/res/text/de-de/normalize.json:55]: actually, this was a mindless copy paste from the old mycroft implementation. thx

**Comment by emphasize** [comment]
But found some edge case while porting this to ovos-classifiers that could be handled in the meantime

**Pull request merged**